runtime/src/launch/slurm-srun/launch-slurm-srun.c

```diff
@@ -89,18 +90,11 @@ static sbatchVersion determineSlurmVersion(void) {
   }
 }
 
-static int nomultithread(int batch) {
```

Comment:
Don't think we can get rid of the nomultithread logic. This was added in https://github.com/chapel-lang/chapel/pull/15176 to address a user issue where a site had `SLURM_HINT=nomultithread` set and if we tried to ask for the logical cores it would fail to allocate nodes. See that PR and the linked issues for more details.

Reply:
The situation is rather complicated, as are all things related to `slurm`.  Keeping the `--hint=nomultithread` functionality as-is won't result in the desired behavior. If one runs 2 locales on an EX node with 2 sockets, 64 cores/socket, 2 PUs/core, and `--hint=nomultithread`, one would expect that each locale would run on its own socket using 64 PUs on 64 cores.  Instead, what's likely to happen with the current launcher implementation (at least under some `slurm` configurations) is that each locale will be bound to the same 128 PUs on the 32 cores in each socket (i.e., half of the cores in each socket go unused, the locales share PUs, and each locale spans both sockets). This is obviously undesirable. Here are some relevant things that cause this to happen. (I use the term `slurm` below to refer to both `srun` and `sbatch`).

* Note that the `--cpus-per-task` `slurm` option refers to PUs if hyperthreading is  enabled and cores if it is not.
* The Chapel `slurm-srun` launcher passes the `--exclusive` option to `slurm` so that Chapel jobs have exclusive access to nodes. This option also causes co-located tasks of a job use disjoint PUs (tasks don't share a PU). 

* The `slurm` `--cpu-bind=none` option doesn't mean that there is no CPU-binding under any circumstance, it means that the individual tasks of a job are not bound to disjoint PUs. If the tasks do not consume all PUs then they will be bound to their subset of the PUs. If `--cpu-bind` is not specified then each task will be bound to disjoint PUs. Here is a table to illustrate. There are 2 tasks (locales) on an EX and the `--exclusive `option is specified. Note that this is under a particular `slurm` configuration, other configurations may behave differently.

| --cpus-per-task | --cpu-bind | Result                                                         |
|-----------------|------------|----------------------------------------------------------------|
|             128 | N/A        | Each task bound to 128 PUs in its own socket                   |
|              64 | N/A        | Each task bound to 64 PUs on 32 cores in its own socket        |
|             128 | none       | No CPU binding at all                                          |
|              64 | none       | Each task bound to the same 128 PUs on 32 cores in each socket |

* In the first two rows `--cpu-bind` is not specified and `slurm` binds the tasks to disjoint PUs. If `--cpu-bind=none` then `slurm` either doesn't bind the tasks at all if `--cpus-per-task=128` because the tasks use all PUs on the node (row 3), or if `--cpus-per-task=64` it allocates a total of 128 PUs for the 2 tasks and binds both tasks to all 128 PUs (row 4).

* The `--hint=nomultithread` option tells `slurm` not to use `hyperthreading`  (i.e. only bind tasks to one PU per core). Note that the `--hint` and `--cpu-bind` options are mutually exclusive; if both are specified `slurm` ignores the `--hint` option (the `SLURM_HINT` environment variable has the same behavior except that `slurm` doesn't print a warning about the hint being ignored). This makes sense because` --hint=nomultithread` tells `slurm` not to bind tasks to more than one PU per core, while `--cpu-bind=none` tells `slurm` not to bind the tasks at all. The latter takes precedence). Here is the same table as above with `--hint=nomultithread` specified.

| --cpus-per-task | --cpu-bind | Result                                                                        |
|-----------------|------------|-------------------------------------------------------------------------------|
|             128 | N/A        | error: Requested node configuration is not available                          |
|              64 | N/A        | Each task bound to 64 cores in its own socket                                 |
|             128 | none       | No CPU binding at all (hint ignored)                                          |
|              64 | none       | Each task bound to the same 128 PUs on 32 cores in each socket (hint ignored) |

* Row 1 is an error because `slurm` is asked to run 2 tasks on 128 disjoint PUs without using hyperthreading, which is impossible.

* The current implementation of the Chapel `slurm-srun` launcher does not  specify `--cpu-bind=none` and if `--hint=nomultithread` it divides the number of CPUs per task by the number of PUs/core to calculate `--cpus-per-task`. This corresponds to row 2, so that each task is bound to the cores in its own socket (note again that different slurm configurations may behave differently).

* There is at least one customer system in which there is a system-wide  `SLURM_HINT=nomultithread` to essentially disable hyperthreading.

* The locale-per-socket functionality requires` --cpu-bind=none` so that each locale can limit itself to its own socket.  However, `--cpu-bind=none` will cause slurm to ignore `--hint=nomultithread`. As currently implemented,  the launcher will still divide by the number of PUs/core to end up with `--cpus-per-task=64` (row 4 above) which is not the outcome we want.

* The reason for specifying `--hint=nomultithread` is to prevent the use  of hyperthreading. Note that by default the Chapel runtime does not use hyperthreading; it creates one qthread shepherd per core. So it does the "right thing" whether or not `--hint=nomultithread` is specified. This corresponds to row 3 above in which `--cpus-per-task=128`, `--cpu-bind=none`, and `--hint=nomultithread` is ignored.

* The user can force Chapel to use hyperthreading by setting `CHPL_RT_NUM_THREADS_PER_LOCALE` to a value greater than the number of cores.

* Base on all of these considerations, if `--hint=nomultithread` the proper behavior by the launcher is to:

1.  **not** reduce `--cpus-per-task`, and
2. limit `CHPL_RT_NUM_THREADS_PER_LOCALE` to the number of cores so that the Chapel runtime does not use hyperthreading.

Reply:
I've done a few experiments since the above comment and have an alternative solution:

1. *Not* reduce `--cpus-per-task`, same as above.
2. Always specify `--cpu-bind=none` even if there is one locale per node. If you don't specify this then #1 will cause an error because `srun` can't bind the locale to that many cores.
3. If `--hint=nomultithread` have the launcher set an environment variable (e.g. `CHPL_RT_NO_MULTITHREAD`). The runtime uses this in `setupAvailableParallelism` in `tasks-qthreads.c` to determine whether to limit the available parallelism to logical or physical CPUs.

Reply:
I implemented a slightly different alternative from above that doesn't change the behavior when there is a single locale per node.

1. Only reduce `--cpus-per-task` if there is a single locale per node.
2. Only specify `--cpu-bind=none` if there is more than one locale per node.
3. If there is more than one locale per node and `--hint=nomultithread`, the launcher sets the environment variable `CHPL_RT_NO_MULTITHREAD` that the runtime uses to determine whether to limit the available parallelism to logical or physical CPUs.